File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        result = [0]* len(temperatures)
        stack = []
        for i,t in enumerate(temperatures):
            while stack and temperatures[stack[-1]] < t:
                index = stack.pop()
                result[index] = i - index 
            stack.append(i)
        return result

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(std::vector<int>& temperatures) {
        vector<int> result(temperatures.size(), 0);
        vector<int> stack;
        stack.reserve(temperatures.size());
        for (int i = 0; i < temperatures.size(); i++) {
            while (!stack.empty() && temperatures[stack.back()] < temperatures[i]) {
                result[stack.back()] = i - stack.back();
                stack.pop_back();
            }
            stack.push_back(i);
        }
        return result;
    }
};
//time complexity = O(n) 
//space complexity = O(n)
